### PR TITLE
UI changes of Metric page

### PR DESCRIPTION
**Pull request by KD1712**
Replaced ID column with Context column, added context filter option. In ADD METRIC json-editor, added context list and removed highlighted blue line

### CODE REVIEW

**Comment by KD1712** [comment]
Screenshot 2024-05-21 at 10 46 23 AM

@zackcl  We are updating the UI of our table component, which includes a hierarchical tree structure. Initially, we used drop-down icons for all nodes. Now, changes made here to create difference between leaf node and expandable are: Expandable nodes use the drop-down icon and leaf nodes use bullet points (•).

Which design option do we prefer for representing leaf nodes in our table component:
**Bullet Points** OR **Empty Space** OR **Disabled drop-down icon** ?

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.html [hunk 1]
```diff
@@ -32,64 +43,58 @@
...
+                    <mat-icon class="mat-icon-rtl-mirror">
+                      {{ nestedTreeControl.isExpanded(node) ? 'expand_more' : 'chevron_right' }}
+                    </mat-icon>
+                  </button>
+                  {{ node.key }}
+                  <mat-icon *ngIf="permissions.metrics.delete && keyEditMode" class="icon" (click)="deleteNode(node)"
+                    >delete_outline</mat-icon
+                  >
+                </div>
+                <div [class.example-tree-invisible]="!nestedTreeControl.isExpanded(node)" role="group">
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.html:93]: replace the name of the class `.example-tree` with the relevant name in the HTML and CSS file.

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.scss [hunk 2]
```diff
@@ -5,6 +5,40 @@
   padding: 0;
   height: 100%;
 
+  .example-tree-invisible {
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.scss:8]: here as well.

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.scss [hunk 3]
```diff
@@ -5,6 +5,40 @@
...
+    display: none;
+  }
+
+  .example-tree ul,
+  .example-tree li {
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.scss:13]: here too.

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.scss [hunk 4]
```diff
@@ -5,6 +5,40 @@
...
+    background-color: black; /* Color of the square */
+  }
+  .example-tree .mat-nested-tree-node div[role='group'] {
+    padding-left: 40px;
+  }
+  .example-tree div[role='group'] > .mat-tree-node {
+    padding-left: 40px;
+  }
+
+  .example-tree-invisible {
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.scss:38]: in these three lines too

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts [hunk 5]
```diff
@@ -18,6 +18,8 @@ import { AnalysisService } from '../../../../../core/analysis/analysis.service';
 import { AuthService } from '../../../../../core/auth/auth.service';
 import { DeleteMetricsComponent } from '../modals/delete-metrics/delete-metrics.component';
 import { AddMetricsComponent } from '../modals/add-metrics/add-metrics.component';
+import { EXPERIMENT_SEARCH_KEY } from '../../../../../core/experiments/store/experiments.model';
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts:21]: Always import enums from `upgrade_types` when possible.

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts [hunk 6]
```diff
@@ -40,12 +42,16 @@ export class MetricsComponent implements OnInit, OnDestroy, AfterViewInit {
...
-  nestedTreeControl: NestedTreeControl<MetricUnit>;
-  nestedDataSource: MatTreeNestedDataSource<MetricUnit>;
+  // _dataChange = new BehaviorSubject<MetricUnit[]>([]);
+  nestedTreeControl = new NestedTreeControl<IMetricUnit>((node) => node.children);
+  nestedDataSource = new MatTreeNestedDataSource<IMetricUnit>();
   insertNodeIndex = 0;
 
   selectedMetricIndex = null;
+  experimentFilterOptions = [EXPERIMENT_SEARCH_KEY.ALL, EXPERIMENT_SEARCH_KEY.NAME, EXPERIMENT_SEARCH_KEY.CONTEXT];
+  selectedExperimentFilterOption = EXPERIMENT_SEARCH_KEY.ALL;
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts:52]: update the name of the variable copied from the experiment here.

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts [hunk 7]
```diff
@@ -57,20 +63,22 @@ export class MetricsComponent implements OnInit, OnDestroy, AfterViewInit {
     this.allMetricsSub = this.analysisService.allMetrics$.subscribe((metrics) => {
       this.allMetrics = new MatTableDataSource();
       this.allMetrics.data = metrics;
+      this.nestedDataSource.data = this.allMetrics.data;
+      this.extractContext(this.allMetrics.data);
     });
+    this.applyFilter(this.selectedExperimentFilterOption);
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts:69]: here as well

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts [hunk 8]
```diff
@@ -57,20 +63,22 @@ export class MetricsComponent implements OnInit, OnDestroy, AfterViewInit {
...
 
-    this.nestedTreeControl = new NestedTreeControl<MetricUnit>(this._getChildren);
-    this.nestedDataSource = new MatTreeNestedDataSource();
-
-    this._dataChange.subscribe((data) => (this.nestedDataSource.data = data));
+    // this.nestedTreeControl = new NestedTreeControl<IMetricUnit>((node) => node.children);
+    // this.nestedDataSource = new MatTreeNestedDataSource<IMetricUnit>();
+    // this._dataChange.subscribe((data) => (this.nestedDataSource.data = data));
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts:73]: removed commented code

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts [hunk 9]
```diff
@@ -112,17 +120,38 @@ export class MetricsComponent implements OnInit, OnDestroy, AfterViewInit {
...
 
-  setTreeForMetric(index: number) {
-    this.selectedMetricIndex = index;
-    this.insertNodeIndex = 0;
-    this.createTree([this.allMetrics.data[index]]);
+  // setTreeForMetric(index: number) {
+  //   this.selectedMetricIndex = index;
+  //   this.insertNodeIndex = 0;
+  //   this.createTree([this.allMetrics.data[index]]);
+  // }
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts:127]: remove commented code

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts [hunk 10]
```diff
@@ -112,17 +120,38 @@ export class MetricsComponent implements OnInit, OnDestroy, AfterViewInit {
...
 
   applyFilter(filterValue: string) {
-    this.selectedMetricIndex = null;
-    this.analysisService.setMetricsFilterValue(filterValue);
+    this.filterExperimentPredicate(this.selectedExperimentFilterOption);
+    if (typeof filterValue === 'string') {
+      this.allMetrics.filter = filterValue.trim().toLowerCase();
+    }
+  }
+  filterExperimentPredicate(type: EXPERIMENT_SEARCH_KEY) {
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts:139]: replace experiment with metrics in all variables in this func.

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts [hunk 11]
```diff
@@ -139,5 +168,22 @@ export class MetricsComponent implements OnInit, OnDestroy, AfterViewInit {
...
+    const contextValuesSet = new Set<string>(); // Using Set to automatically remove duplicates
+
+    function extractContextRecursively(obj: any) {
+      if (obj.context && Array.isArray(obj.context)) {
+        obj.context.forEach((value: string) => contextValuesSet.add(value));
+      }
+      if (obj.children && Array.isArray(obj.children)) {
+        obj.children.forEach((child: any) => extractContextRecursively(child));
+      }
+    }
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts:181]: move this func outside the `extractContext` func.

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts [hunk 12]
```diff
@@ -11,10 +13,18 @@ import { AnalysisService } from '../../../../../../core/analysis/analysis.servic
 export class AddMetricsComponent implements OnInit {
   options = new JsonEditorOptions();
   metricsEditorError = false;
+  allMetrics: any;
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts:16]: replace any with type.

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts [hunk 13]
```diff
@@ -11,10 +13,18 @@ import { AnalysisService } from '../../../../../../core/analysis/analysis.servic
...
+  allMetricsSub: Subscription;
+  contextOptions: string[] = [];
   @ViewChild('metricsEditor', { static: false }) metricsEditor: JsonEditorComponent;
   constructor(private dialogRef: MatDialogRef<AddMetricsComponent>, private analysisService: AnalysisService) {}
 
   ngOnInit() {
+    this.allMetricsSub = this.analysisService.allMetrics$.subscribe((metrics) => {
+      this.allMetrics = new MatTableDataSource();
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts:24]: `MatTableDataSource` is for the table data source, shouldn't be used here.

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts [hunk 14]
```diff
@@ -48,4 +58,20 @@ export class AddMetricsComponent implements OnInit {
...
+    const contextValuesSet = new Set<string>(); // Using Set to automatically remove duplicates
+
+    function extractContextRecursively(obj: any) {
+      if (obj.context && Array.isArray(obj.context)) {
+        obj.context.forEach((value: string) => contextValuesSet.add(value));
+      }
+      if (obj.children && Array.isArray(obj.children)) {
+        obj.children.forEach((child: any) => extractContextRecursively(child));
+      }
+    }
```
Comment (Yagnik56) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts:71]: move func outside the parent func.

**Review by Yagnik56 — COMMENTED** [review]
please look into comments I added

**Comment by KD1712** [comment]
> please look into comments I added

resolved all comments

**Pull request closed without merging**